[Imran] iP

Imranr2 opened the pull request:
## Lifeline
> “Your mind is for having ideas, not holding them.” – David Allen [(source)](https://dansilvestre.com/productivity-quotes/)

Lifeline frees your mind of having to remember things you need to do. It's,
* text-based
* easy to learn
* ~~FAST~~ *SUPER* FAST to use

All you need to do is,
1. download it from [here](https://github.com/Imranr2/ip/releases).
2. run the jar using the command line.
3. add your tasks.
4. let it manage your tasks for you :smile:

Most importantly it is **FREE**!

Features:
- [x] Managing tasks
- [x] Managing deadlines
- [x] Managing events
- [x] GUI

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Lifeline lifeline = new Lifeline("save" + File.separator + "tasks.json");
        lifeline.start();
    }
}
```

Commit:
Add Gradle support

Imranr2 committed:
feat: Changed default logo to "Lifeline"

Imranr2 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

Imranr2 committed:
chore: Added checkstyle files

Imranr2 committed:
refactor: Rename Duke.java to Lifeline.java

Imranr2 committed:
refactor: Uncomment checkstyle

Imranr2 committed:
feat: Add Greet, Echo, Exit to Lifeline

Imranr2 committed:
feat: Add level-2 features (Add, List)

Imranr2 committed:
feat: Add Task class to represent tasks

Imranr2 committed:
feat: Add capability to mark tasks as done

Imranr2 committed:
refactor: Shift Lifeline main function to Main.java

Imranr2 committed:
feat: Add ToDo class

This class extends from Task

Imranr2 committed:
feat: Add Deadline class

This class extends from Task and allows a task with a deadline.

Imranr2 committed:
feat: Add Event class

This class extends from Task and allows for an event with a duration.

Imranr2 committed:
feat: Add level-4 features (ToDo, Event, Deadline)

Imranr2 committed:
feat: List function now shows number of uncompleted tasks

Imranr2 committed:
style: Add line breaks to long lines

Imranr2 committed:
test: Add Automated Text UI Testing

Imranr2 committed:
fix: Remove redundant full stop for done output

Imranr2 committed:
test: Update EXPECTED.txt

Imranr2 committed:
fix: Change accessor to private

Imranr2 committed:
style: Add newline to end of file

Imranr2 committed:
style: Move colon to next line

Imranr2 committed:
refactor: Uncomment checkstyle

Imranr2 committed:
feat: Add custom exception for Lifeline

Imranr2 committed:
feat: Add LifelineException to handle errors
Cases handled by LifelineException:
* No description for todo/deadline/event
* No deadline for deadline task
* No timeframe for event
* Index out of bounds when marking task as done
* Non-integer index when marking task as done
* No integer when marking task as done

Imranr2 committed:
test: Update input.txt and EXPECTED.txt to test error handling

Imranr2 committed:
feat: Add level-6 features (Delete)

Imranr2 committed:
test: Update input.txt and EXPECTED.txt to test delete function

Imranr2 committed:
feat: Add enums to better handle user input

Imranr2 committed:
chore: Add Override tag to ToString method

Imranr2 committed:
refactor: Use enums to handle user input

Imranr2 committed:
test: Update EXPECTED.TXT

Imranr2 committed:
refactor: Trim user input

Imranr2 committed:
feat: Add Gson library to save tasks as JSON

Imranr2 committed:
refactor: Add new constructors to Task and subclasses

New constructor used to support load function

Imranr2 committed:
refactor: Add save directory to store saved tasks

Imranr2 committed:
feat: Add level-7 features (Save)

Users tasks are now saved to ./save/tasks.json. When user starts the chatbot again, load function will read the json file.

Imranr2 committed:
Merge branch 'branch-level-7'

Imranr2 committed:
chore: Add newline to end of file

Imranr2 committed:
feat: Add level-8 features (Dates and Times)
* Event now has LocalDate to represent date of event, LocalTime to
represent start time and end time.
* Date now has LocalDateTime to represent deadline date and time

Imranr2 committed:
Merge branch 'branch-level-8'

Imranr2 committed:
feat: Add Tasklist to manage tasks

Imranr2 committed:
feat: Add Storage to handle saving and loading of tasks

Imranr2 committed:
feat: Incorporate Storage and TaskList into Lifeline

Imranr2 committed:
fix: Missing word

Imranr2 committed:
feat: Add Ui to handle interaction with user

Imranr2 committed:
feat: Incorporate Ui into Lifeline
Ui now handles all ui interactions

Imranr2 committed:
refactor: Rename InputType to Command

Imranr2 committed:
feat: Add Parser to parse user input
Parser returns a command that is used in the switch case

Imranr2 committed:
fix: Update toString method to show h:mm

Imranr2 committed:
fix: Tasks not saving after being marked as completed

Imranr2 committed:
feat: Add CommandHandler to replace switch statement

Imranr2 committed:
feat: Add QuadFunction for functional enums

Imranr2 committed:
fix: Update Quadfunction to return void

Imranr2 committed:
fix: Update displayed messages

Imranr2 committed:
refactor: Huge refactor to Lifeline
* Handling of different commands through CommandHandler. Each
command has their own handler
* Command enums are now functional, each enum has a CommandHandler
function
* Saving and loading of tasks controlled by Storage
* Interaction with Ui through Ui class
* Parsing of user command through Parser class
* Tasks are now managed by TaskList instead of standalone Arraylist
* Minor ui fixes (chatbot displayed done instead of deleted
when deleting task)

Imranr2 committed:
refactor: Divide classes into packages

Imranr2 committed:
chore: Rearrange imports

Imranr2 committed:
refactor: Shift Main and Lifeline into lifeline package

Imranr2 committed:
fix: Create directory to save tasks if directory is missing

Imranr2 committed:
refactor: Change error message

Imranr2 committed:
fix: Create directory not working when no folder is specified

Imranr2 committed:
refactor: UI and CommandHandler now return strings
Change was done to prepare for implementation of JavaFx
UI now has a function printToConsole
Lifeline and Quadfunction reworked to accommodate changes

Imranr2 committed:
fix: Close file reader after reading file during load

Imranr2 committed:
style: Fix checkstyle

Imranr2 committed:
test: Add JUnit tests for ToDo

Imranr2 committed:
test: Add JUnit tests for Task

Imranr2 committed:
test: Add JUnit tests for TaskList

Imranr2 committed:
test: Add JUnit tests for Event

Imranr2 committed:
test: Add JUnit tests for Deadline

Imranr2 committed:
test: Add JUnit tests for Storage

Imranr2 committed:
test: Add JUnit tests for Parser

Imranr2 committed:
test: Add JUnit tests for LifelineException

Imranr2 committed:
test: Add JUnit tests for CommandHandler

Imranr2 committed:
fix: Tick and cross symbols not displaying correctly in terminal

Imranr2 committed:
fix: Sad face not showing correctly in terminal

Imranr2 committed:
fix: Remove unicode characters and update test

Imranr2 committed:
docs: Add Javadoc for Lifeline

Imranr2 committed:
docs: Add Javadoc for Main

Imranr2 committed:
docs: Add Javadoc for CommandHandler

Imranr2 committed:
docs: Add Javadoc for LifelineException

Imranr2 committed:
docs: Add Javadoc for Parser

Imranr2 committed:
docs: Add Javadoc for Storage

Imranr2 committed:
docs: Add Javadoc for Deadline

Imranr2 committed:
docs: Add Javadoc for Event

Imranr2 committed:
docs: Add Javadoc for Task

Imranr2 committed:
docs: Add Javadoc for TaskList

Imranr2 committed:
docs: Add Javadoc for ToDo

Imranr2 committed:
docs: Add Javadoc for ui

Imranr2 committed:
chore: Remove unused import

Imranr2 committed:
chore: Add comments and breaks for better readability

Imranr2 committed:
feat: Add support to find task given a single keyword

Imranr2 committed:
test: Add test for find command

Imranr2 committed:
refactor: Change output message for finding tasks

Imranr2 committed:
feat: Change find command to be case-insensitive

Imranr2 committed:
Merge branch 'branch-A-Javadoc'

Imranr2 committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/lifeline/command/CommandHandler.java

Imranr2 committed:
Merge branch 'branch-level-9'

# Conflicts:
#	src/main/java/lifeline/command/CommandHandler.java
#	src/main/java/lifeline/ui/Ui.java

Imranr2 committed:
docs: Update Javadoc for CommandHandler

Imranr2 committed:
docs: Update Javadoc for Ui

Imranr2 committed:
style: Fix indentation and new lines to comply with checkstyle

Imranr2 committed:
docs: Remove redundant Javadocs

Imranr2 committed:
refactor: Move addition of tasks to TaskList private function

Imranr2 committed:
style: Improve code quality

Imranr2 committed:
fix: Catch out of bounds error for delete and done commands

Imranr2 committed:
test: Add tests to test for invalid index

Imranr2 committed:
feat: Add JavaFX dependency

Imranr2 committed:
feat: Add images for GUI

Imranr2 committed:
feat: Add basic GUI
GUI implemented using JavaFX
Styling yet to be added

Imranr2 committed:
feat: Add styling to GUI

Imranr2 committed:
feat: Allow user to choose between console or GUI

Imranr2 committed:
Merge branch 'branch-GUI'

Imranr2 committed:
feat: Change window icon and chat images

Imranr2 committed:
feat: Change font and window size

Imranr2 committed:
feat: Disable window resizing

Imranr2 committed:
feat: Add welcome message when user starts GUI

Imranr2 committed:
docs: Update JavaDocs

Imranr2 committed:
refactor: TaskList should handle finding of tasks

Imranr2 committed:
feat: Add help message to guide new users

Imranr2 committed:
docs: Update class header comments

Imranr2 committed:
feat: Add alias for commands

Imranr2 committed:
Merge branch 'branch-A-Varargs'

Tanishq4331 commented on src/main/java/lifeline/Main.java:
```diff
@@ -0,0 +1,39 @@
+        try {
+            FXMLLoader fxmlLoader = new FXMLLoader(Main.class.getResource("/view/MainWindow.fxml"));
+            AnchorPane ap = fxmlLoader.load();
+            Scene scene = new Scene(ap);
```
Perhaps the variable  `ap` could be renamed to `anchorPane` to make it more descriptive.

Imranr2 replied:
Thank you for pointing it out. I will make the changes

Tanishq4331 commented on src/main/java/lifeline/task/TaskList.java:
```diff
@@ -0,0 +1,108 @@
+     *
+     * @return Number of tasks in TaskList
+     */
+    public int size() {
```
This method name could be further improved to use a verb instead. Maybe something like `getListSize`?

Imranr2 replied:
I was trying to follow the arraylist convention but you're right getListSize is more descriptive.

Tanishq4331 commented on src/main/java/lifeline/storage/Storage.java:
```diff
@@ -0,0 +1,126 @@
+    /**
+     * File path to save the tasks
+     */
+    private String filepath;
```
Perhaps the field `filepath` could be renamed to `filePath` to follow camelCase convention.

Imranr2 replied:
I must've missed this. Thanks for pointing it out

Tanishq4331 commented on src/main/java/lifeline/ui/Ui.java:
```diff
@@ -0,0 +1,198 @@
+ */
+public class Ui {
+
+    private Scanner sc;
```
Maybe you could consider renaming `sc` to something more descriptive.

Imranr2 replied:
You're right I should rename it to scanner instead. Thanks

Tanishq4331 reviewed (commented):
The code does well to meet the code quality guidelines. There are just a few places where you could reconsider variable names.

Imranr2 committed:
refactor: Rename some parts of code to match coding standard

Imranr2 committed:
fix: Change font to ensure compatibility with all devices

Shruthi0310 commented on src/main/java/lifeline/Lifeline.java:
```diff
@@ -0,0 +1,102 @@
+package lifeline;
+
+import java.util.ArrayList;
+
+import lifeline.command.Command;
+import lifeline.exception.LifelineException;
+import lifeline.parser.Parser;
+import lifeline.storage.Storage;
+import lifeline.task.TaskList;
+import lifeline.ui.Ui;
```
I like how you have separated these imports to make it more organized and readable!

Imranr2 replied:
Thank you!

Shruthi0310 commented on src/main/java/lifeline/Lifeline.java:
```diff
@@ -0,0 +1,102 @@
+     *
+     * @param filepath Path to load or save tasks
+     */
+    public Lifeline(String filepath) {
```
Should filepath be in camelCase instead (filePath)?

Imranr2 replied:
Yup you are right. I have fixed it in the latest version. Thanks for pointing it out!

Shruthi0310 commented on src/main/java/lifeline/ui/Ui.java:
```diff
@@ -0,0 +1,198 @@
+                    uncompletedTask++;
+                }
+            }
+            stringBuilder.append("You have " + uncompletedTask + " uncompleted " + (uncompletedTask > 1 ? "tasks"
+                    : "task") + ".\n");
```
Perhaps you could put the conditional on the same line or split into 3 different lines to follow the coding standards.

Imranr2 replied:
Thanks for pointing it out!

Shruthi0310 commented on src/test/java/lifeline/command/CommandHandlerTest.java:
```diff
@@ -0,0 +1,321 @@
+import lifeline.task.ToDo;
+import lifeline.ui.Ui;
+
+public class CommandHandlerTest {
```
I like the way you named the tests as it makes it easier to know which method is tested and the test scenario, which follows the coding standard well.

Imranr2 replied:
Thank you!

Shruthi0310 commented on src/main/java/lifeline/Lifeline.java:
```diff
@@ -0,0 +1,102 @@
+     * Gets the appropriate response from Lifeline if user inputs a valid command
+     * Method is used only for GUI
+     *
+     * @param command User input
```
Perhaps could put punctuation behind the parameter descriptions as mentioned in the coding standards doc.

Imranr2 replied:
I completely missed that. Thanks!

Shruthi0310 commented on src/main/java/lifeline/task/TaskList.java:
```diff
@@ -0,0 +1,108 @@
+     * @param index Index of task to be returned
+     * @return Task at specified index
+     */
+    public Task get(int index) {
```
Maybe could use the name getTask rather than get to better following the naming convention.

Imranr2 replied:
I was trying to follow the arrayList convention but you're right it should be named getTask. Thanks!

Shruthi0310 commented on src/main/java/lifeline/Lifeline.java:
```diff
@@ -0,0 +1,102 @@
+    }
+
+    private void getInput() {
+        boolean exit = false;
```
Should this be isExit instead since its a boolean value?

Imranr2 replied:
Good catch! Thanks for pointing it out

Shruthi0310 commented on src/main/java/lifeline/task/Task.java:
```diff
@@ -0,0 +1,52 @@
+        this.isDone = isDone;
+    }
+
+    public void setDone(boolean done) {
```
Same issue for naming convention for boolean as mentioned earlier.

Shruthi0310 reviewed (commented):
Hey, overall, I feel that your code followed the coding standards very well and was very easy to read as you split the codes at necessary places! I didn't find any major coding standard violations just some minor ones 😃

Imranr2 committed:
style: Update code to comply with java coding standard
Add period to end of JavaDocs
Rename boolean parameters

IrvingHe000 commented on src/main/java/lifeline/Launcher.java:
```diff
@@ -0,0 +1,48 @@
+     *
+     * @param args Arguments specified by the user.
+     */
+    public static void main(String[] args) {
+        try {
+            String mode = getMode(args);
+            if (mode.equals("console")) {
+                startConsole();
+            } else {
+                Application.launch(Main.class, args);
+            }
+        } catch (ArrayIndexOutOfBoundsException e) {
+            System.out.println("You did not specify a mode. Please specify 'gui' or 'console'");
+        }
+    }
```
I like your way that providing two ways to start the program :)

IrvingHe000 commented on src/main/java/lifeline/task/Deadline.java:
```diff
@@ -0,0 +1,41 @@
+ */
+public class Deadline extends Task {
+
+    private LocalDateTime by;
```
Perhaps can change the name to a more meaningful one?

IrvingHe000 commented on src/main/java/lifeline/task/Task.java:
```diff
@@ -0,0 +1,52 @@
+
+    @Override
+    public String toString() {
+        if (isDone) {
+            return "[X] " + this.name;
+        } else {
+            return "[ ] " + this.name;
+        }
```
```suggestion
        String isDoneMarker = isDone ? "[X] " : "[ ] ";
        return isDoneMarker + this.name;
```

IrvingHe000 commented on src/main/java/lifeline/ui/Ui.java:
```diff
@@ -0,0 +1,198 @@
+            stringBuilder.append("You have no remaining tasks.\n");
+        } else {
+            int uncompletedTask = 0;
+            stringBuilder.append("Here " + (taskList.getSize() > 1 ? "are" : "is")
+                    + " your " + (taskList.getSize() > 1 ? "tasks:" : "task:") + "\n");
```
Perhaps using append instead of '+' will be more efficient.

IrvingHe000 reviewed (commented):
Overall it is very neat code and easy to read :) 👍

Imranr2 committed:
feat: Add Java Assertions to CommandHandler

CommandHandler does not have assertions to check for if command
passed to methods are of the correct type

The lack of assertions make it possible for a command to be passed
to the wrong method

Add assertions to ensure that command passed to the method
is of the correct type
E.g. handleDeadline should only handle deadline commands

Assertions allow for checking of programmer mistakes

Imranr2 committed:
test: Supply correct arguments to test cases

Imranr2 committed:
Merge branch 'master' into branch-A-Assertions

Imranr2 committed:
feat: Add Java Assertions to TaskList
TaskList does not have assertions to check if the index passed
to the methods deleteTask, getTask and completeTask is not out
of bounds

The lack of assertions make it possible for an IndexOutOfBounds
exception to be thrown

Add assertions to ensure that index to be passed to TaskList is
within the bound

Assertions allow for checking of programmer mistakes

Imranr2 committed:
Merge pull request #1 from Imranr2/branch-A-Assertions

A-Assertions

Imranr2 committed:
refactor: Use named constants to represent errors in CommandHandler
Test update to reflect changes

Imranr2 committed:
feat: Add more error strings

Imranr2 committed:
refactor: Replace error message with ErrorStrings

Imranr2 committed:
style: Rearrange imports

Imranr2 committed:
refactor: Update Storage to use ErrorStrings

Imranr2 committed:
style: Shift operator to next line to comply with checkstyle

Imranr2 committed:
Merge pull request #2 from Imranr2/branch-A-CodeQuality

A-CodeQuality

Imranr2 committed:
fix: Fix unable to parse command with leading spaces

Imranr2 committed:
fix: Command not recognised due to leading and trailing spaces

Imranr2 committed:
refactor: Extract out methods into smaller methods

Imranr2 committed:
feat: Allow tasklist to delete and complete multiple tasks

Imranr2 committed:
feat: Allow user to delete and complete multiple tasks at once
Users are only able to delete and complete single tasks.

This is inconvenient for users who want to perform operations on
multiple tasks.

Add functionality for tasklist to support mass operations
Change CommandHandler to handle multiple indices

Allows user to input comma separated indices to perform mass
operations such as deleting multiple tasks at once

Imranr2 committed:
docs: Add JavaDocs to public methods

Imranr2 committed:
Merge branch 'branch-C-MassOps'

Imranr2 committed:
refactor: Huge refactor to comply with SLAP

Imranr2 committed:
docs: Add JavaDocs to public method

Imranr2 committed:
feat: Add assertion to ensure command is of correct type

Imranr2 committed:
refactor: Remove unused methods

Imranr2 committed:
style: Rearrange lines

Imranr2 committed:
fix: Storage not loading Deadline task correctly

Imranr2 committed:
feat: Catch alias command with empty details

Imranr2 committed:
feat: Add alias command for users to view available aliases

Imranr2 committed:
refactor: Remove redundant method

Imranr2 committed:
refactor: Extract filepath as local variable

Imranr2 committed:
feat: Update GUI response message

Imranr2 committed:
feat: Update GUI response message

Imranr2 committed:
feat: Add alias rm to delete

Imranr2 committed:
docs: Update Readme.md

Imranr2 committed:
docs: Update image path in README.md

Imranr2 committed:
docs: Update README.md images

Imranr2 committed:
Set theme jekyll-theme-cayman

Imranr2 committed:
docs: Update image tags in README.md

Imranr2 committed:
docs: Fix spelling issues

Imranr2 committed:
Set theme jekyll-theme-midnight

Imranr2 committed:
Set theme jekyll-theme-hacker

Imranr2 committed:
refactor: Rearrange code for better readability

Imranr2 committed:
fix: Catch missing description for deadline and event

Imranr2 committed:
fix: Fix event error not displaying correct error message

Imranr2 committed:
fix: Validate first word of command for list,bye,help,alias

Imranr2 committed:
fix: Handle empty comma separated string

Imranr2 committed:
Update README.md

Imranr2 committed:
Update README.md

Imranr2 committed:
feat: Display error if end time is after start time for events

Imranr2 committed:
Update README.md to include acknowledgements

Imranr2 committed:
feat: Use free to reuse images for lifeline

Imranr2 committed:
Merge branch 'master' of github.com:Imranr2/ip

Imranr2 committed:
docs: Update images to reflect current UI
